e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -233,24 +223,77 @@ func (pf *PortForwarder) start() error {
 	// Restart the port forward if the process exits prematurely (e.g. due to pod eviction)
 	go func() {
 		for {
-			if err := pf.watcher.WatchForNotFound(kinds.Pod(), pf.pod, pf.ns, testwatcher.WatchTimeout(0), testwatcher.WatchContext(pf.ctx)); err != nil {
-				if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
-					// stop retrying, the subprocess will be killed with the context.
-					return
-				}
-				// Watch automatically retries, so this shouldn't happen unless
-				// there's a terminal error.
-				pf.logger.Infof("Watch exited, restarting port-forward for %s/%s: %v", pf.ns, pf.deployment, err)
-			}
-			if err := pf.portForwardToDeployment(true); err != nil {
-				pf.logger.Info(err)
-				// this is a potentially fatal error since we failed to restart the port
-				// forward. we log the error and return in case the test is already done
-				// with the port forward. the test will otherwise error the next time it
-				// tries to use the local port.
+			if err := pf.waitAndRestart(); err != nil {
+				pf.logger.Infof("PortForwarder finished: %v", err)
 				return
 			}
 		}
 	}()
 	return nil
 }
+
+// waitAndRestart will block until seeing an indicator that the port-forward is
+// no longer healthy. If that occurs, it will restart the port-forward
+func (pf *PortForwarder) waitAndRestart() error {
+	// create an inner context for the watch so we can cancel threads between iterations
+	// the subprocess will be killed inside portForwardToDeployment
+	watchCtx, watchCancel := context.WithCancel(pf.ctx)
+	defer watchCancel()
+	// restart the port-forward if:
+	// - the subprocess is killed
+	// - the pod is restarted
+	restartChan := pf.signalOnFailure(
+		func() error {
+			<-pf.ctx.Done()
+			return pf.ctx.Err()
+		},
+		func() error {
+			return pf.cmd.Wait()
+		},
+		func() error {
+			return pf.watcher.WatchForNotFound(kinds.Pod(), pf.pod, pf.ns,
+				testwatcher.WatchTimeout(0), testwatcher.WatchContext(watchCtx))
+		},
+	)
+	err := <-restartChan
+	if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+		return fmt.Errorf("exiting - PortForwarder for pod %s/%s received done signal", pf.ns, pf.pod)
+	}
+	pf.logger.Infof("restarting - PortForwarder for pod %s/%s received restart signal", pf.ns, pf.pod)
+	if err := pf.portForwardToDeployment(true); err != nil {
+		// this is a potentially fatal error since we failed to restart the port
+		// forward. we log the error and return in case the test is already done
+		// with the port forward. the test will otherwise error the next time it
+		// tries to use the local port.
+		return err
+	}
+	return nil
+}
+
+// signalOnFailure returns a chan which will be closed if any of the provided
+// functions return. The functions should block and only return if a failure
+// has occurred and the port-forward must be restarted
+func (pf *PortForwarder) signalOnFailure(fns ...func() error) chan error {
```

Comment:
I spent way too long trying to figure understand how this works.
I think it mostly comes down to not using the conventional language.
But when I finally figured it out, I realized there was a better way.

The return channel what I would call a `doneCh`, indicating that the goroutines it's waiting for are done. `restartChan` is confusing, because this the `signalOnFailure` func doesn't handle restarting anything, and it can be closed even if restart is not required (e.g. context cancelled).

With that renamed, it follows that the boolean should be named `closed` or `done`. Yes, it's a signal to the caller, but it's not a signal within the context of this method, because there's no reader here, only writers. Then you'd end up with a channel and boolean with similar names, which is the smell that might have lead discovery of a simpler way, because...

When you're closing a channel, it's generally recommended to only close it on a single goroutine, not with a mutex and locking on each goroutine. This can be accomplished by having the goroutines send a message to an intermediate channel and spawning another goroutine to block waiting for a message on the intermediate channel.

The next question becomes, who should close the channels? The answer is always "the writer". And that's another tip: you should only have one writer for each channel, otherwise there's no way to cleanly close it, because you can't write to a closed channel. From there it follows that we need an intermediate channel for each goroutine, and a dedicated goroutine to close the channel we returned to the caller.

The fan-in aggregation goroutine also needs to wait until all the other goroutines are done, otherwise we could end up with memory leaks.

The nest question is "how do we block waiting for one of the channels in a list of channels?". The answer is the [Select](https://golang.org/pkg/reflect/#Select) function from the [reflect](http://golang.org/pkg/reflect/) package! https://stackoverflow.com/a/19992525 Block waiting for ONE of them to finish, then signal them all to stop (cancel()), then block waiting for ALL of them to finish, then close the `doneCh` you gave to the caller.

Do all that and you can replace the mutex/bool with channels, avoid memory leaks, and make it all easier to read.

Reply:
Oh, the more obvious smell i missed is mixing channels and locks. 

If you're using channels you can probably do it all with channels, but it might require multiple. 
If you're using mutexes, you can probably do it all with a single lock (sync.Mutex or sync.RWMutex or sync.Cond).
Usually go with channels unless there's a good reason you can't.

Reply:
I admittedly have much more experience with multi-threaded programming in C/C++ than Go, so I'm not as familiar with conventions around channels. I'll take a stab at rewriting this with your suggestions in mind

Reply:
Channel-ification done. PTAL